lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -1126,16 +1117,17 @@ private function getFromClauses() : array
 
         // Loop through all FROM clauses
         foreach ($this->sqlParts['from'] as $from) {
-            if ($from['alias'] === null || $from['alias'] === $from['table']) {
-                $tableSql = $from['table'];
+            /** @var From $from */
+            if ($from->alias === null || $from->alias === $from->table) {
+                $tableSql = $from->table;
 
                 /** @var string $tableReference */
-                $tableReference = $from['table'];
+                $tableReference = $from->table;
             } else {
-                $tableSql = $from['table'] . ' ' . $from['alias'];
+                $tableSql = $from->table . ' ' . $from->alias;
 
                 /** @var string $tableReference */
```

Comment:
Cannot remove these `@var` annotations (that were already there), as the actual type is `string|null` and makes phpstan unhappy. I'm not sure what our level of confidence is here in the fact that it will never be `null`?

Reply:
You can assign the alias to a variable before checking whether it's null. This way, in the `else` case, it couldn't be equal to null.

Reply:
Actually in this case, we can be confident that `$table` will never be null, as it can only be created from here:

    public function from(string $from, ?string $alias = null)
    {
        return $this->add('from', new From($from, $alias), true);
    }

So I'd rather separate concerns between `FROM` (with non-nullable `$table`) and `INSERT`/`UPDATE`/`DELETE` as you suggested.

Reply:
Since the separation of concerns is already implemented, is this annotation still needed? As far as I can tell, PHPStan handles the non-null property values correctly within a method: https://phpstan.org/r/700b50cd-9879-4801-9921-f2d5a9c3af3c.

Reply:
The first one was a leftover and not required indeed, but the second still failed with phpstan. Moving the `@var From $from` above the `foreach` fixed it (remembered your previous comment).

Fixed in https://github.com/doctrine/dbal/pull/3833/commits/e99069b5abb081736f187f48d733daea12676895